File: NET/Kata.RockPaperScissor/Shape.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Kata.RockPaperScissor
{
    public abstract class Shape : IShape
    {
        public ShapeType ShapeType { get; protected set; }

        public string Name { get; protected set; }

        protected List<ShapeType> winsTo;

        public bool Defeat(IShape paper)
        {
            return winsTo.Contains(paper.ShapeType);
        }
    }
}

File: JavaScript/src/typescript/Shape.js
var Shape = (function () {
    function Shape(shapeType, winsTo) {
        this.shapeType = shapeType;
        this.winsTo = winsTo;
    }
    Shape.prototype.defeat = function (shapeType) {
        for (var i = 0; i < this.winsTo.length; i++) {
            if (this.winsTo[i] === shapeType) {
                return true;
            }
        }
        return false;
    };
    return Shape;
})();
exports.Shape = Shape;
